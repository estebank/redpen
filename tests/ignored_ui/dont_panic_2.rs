#![allow(redpen::infallible_allocation)]
#![deny(redpen::dont_panic)]
struct S<T: Trait>(T);

struct X;
impl Trait for X {}
trait Trait {
    fn panic() {
        panic!();
    }
}

impl<T: Trait> S<T> {
    fn foo(&self) {
        T::panic();
    }
}

trait A<T: Trait> {
    fn from_trait(&self) {
        T::panic();
    }
    fn from_a_impl(&self);
}
trait B<T> {
    fn from_impl(&self);
}

impl<T: Trait> A<T> for S<T> {
    fn from_a_impl(&self) {
        self.foo();
    }
}
impl<T: Trait> B<T> for S<T> {
    fn from_impl(&self) {
        self.foo();
    }
}

fn impl_fn<T: Trait>(x: impl A<T>) {
    x.from_trait();
    x.from_a_impl();
}
fn trait_object<T: Trait>(x: Box<dyn A<T>>) {
    x.from_trait();
    x.from_a_impl();
}

fn foo1() {
    panic!("shouldn't happen!");
}
fn foo2() {
    S(X).foo();
}
fn foo3() {
    S(X).from_trait();
}
fn foo4() {
    S(X).from_a_impl();
}
fn foo5() {
    S(X).from_impl();
}
fn foo6() {
    (|| {
        panic!();
    })();
}
fn foo7() {
    #[allow(redpen::dont_panic)]
    let x =  || panic!();
    x();
}
fn foo8() {
    let mut x = vec![];
    x.push("");
    x.swap_remove(0);
    let _ = x[1];
}

fn bar() {
    foo1();
}

fn baz() {
}

fn main() {
    foo1();
    foo2();
    foo3();
    foo4();
    foo5();
    foo6();
    foo7();
    bar();
    baz();
}