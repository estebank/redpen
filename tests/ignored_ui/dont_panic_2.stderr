warning: calling `foo8` can unsafe
    --> $DIR/dont_panic_2.rs:76:5
     |
76   |     x.push("");
     |     ^^^^^^^^^^ this can unsafe
     |
note: `std::vec::Vec::<&str>::push` can unsafe here
    --> /home/gh-estebank/.rustup/toolchains/nightly-2024-05-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/vec/mod.rs:1880:13
     |
1876 |     pub fn push(&mut self, value: T) {
     |     --------------------------------
...
1880 |             self.buf.reserve_for_push(self.len);
     |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: `alloc::raw_vec::RawVec::<&str>::reserve_for_push` can unsafe here
    --> /home/gh-estebank/.rustup/toolchains/nightly-2024-05-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/raw_vec.rs:302:24
     |
301  |     pub fn reserve_for_push(&mut self, len: usize) {
     |     ----------------------------------------------
302  |         handle_reserve(self.grow_amortized(len, 1));
     |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: `alloc::raw_vec::RawVec::<&str>::grow_amortized` can unsafe here
    --> /home/gh-estebank/.rustup/toolchains/nightly-2024-05-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/raw_vec.rs:404:19
     |
383  |     fn grow_amortized(&mut self, len: usize, additional: usize) -> Result<(), TryReserveError> {
     |     ------------------------------------------------------------------------------------------
...
404  |         let ptr = finish_grow(new_layout, self.current_memory(), &mut self.alloc)?;
     |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: `alloc::raw_vec::finish_grow::<std::alloc::Global>` can unsafe here
    --> /home/gh-estebank/.rustup/toolchains/nightly-2024-05-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/raw_vec.rs:482:13
     |
464  | / fn finish_grow<A>(
465  | |     new_layout: Result<Layout, LayoutError>,
466  | |     current_memory: Option<(NonNull<u8>, Layout)>,
467  | |     alloc: &mut A,
468  | | ) -> Result<NonNull<[u8]>, TryReserveError>
469  | | where
470  | |     A: Allocator,
     | |_________________-
...
482  |               alloc.grow(ptr, old_layout, new_layout)
     |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
485  |           alloc.allocate(new_layout)
     |           ^^^^^^^^^^^^^^^^^^^^^^^^^^
     = note: `#[warn(redpen::call_unsafe)]` on by default

warning: calling `foo8` can unsafe
   --> $DIR/dont_panic_2.rs:79:1
    |
79  | }
    | ^ this can unsafe
    |
note: `std::ptr::drop_in_place::<std::vec::Vec<&str>>` can unsafe here
   --> /home/gh-estebank/.rustup/toolchains/nightly-2024-05-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:497:1
    |
497 | pub unsafe fn drop_in_place<T: ?Sized>(to_drop: *mut T) {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: 2 warnings emitted

